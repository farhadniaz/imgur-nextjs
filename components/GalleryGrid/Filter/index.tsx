import React, { FC, useState, useCallback } from "react";
import { useSelector } from "react-redux";
import { Row, Col, Button, Modal, Switch, Select } from "antd";
import { FilterOutlined } from "@ant-design/icons";

import StyledDiv, { Vertical } from "./style";
import {
  imgurFilterQuery,
  filterDefautValue,
} from "~/store/gallery/initial-state";
interface Props {
  onFilter: Function;
  loading: boolean;
}

const Filter: FC<Props> = (props) => {
  const { onFilter = () => { }, loading } = props;
  const [modalVisible, setModalVisible] = useState(false);

  const [data, setData] = useState(filterDefautValue);
  const Device = useSelector((state) => state.App.device);
  const applyiedFilter = useSelector((state) => state.Gallery.filter);
  const isThereAnyFilter = JSON.stringify(filterDefautValue) != JSON.stringify(applyiedFilter);
  const isMobieOrTablet =
    Device.isMobieOrTablet || (Device.widthInMobile && Device.isDesktop);


  const setModalInVisible = () => setModalVisible(false);
  const setModalBeVisible = () => setModalVisible(true);

  const doFilter = () => {
    onFilter(data);
    setModalInVisible();
  };
  const restFilter = () => {
    onFilter(filterDefautValue);
    setData(filterDefautValue);
    setModalInVisible();
  };

  const renderOptions = (optionsList) => {
    return Object.keys(optionsList).reverse().map((propertyName) => {
      const item = optionsList[propertyName];
      return (
        <Select.Option key={item} value={item} label={item}>
          {item}
        </Select.Option>
      );
    });
  };

  const filterForm = useCallback((spanSize = 4) => {
    const viralChange = (checked) => setData({ ...data, showViral: checked });
    const windowChange = (value) => setData({ ...data, window: value });
    const sortChange = (value) => setData({ ...data, sort: value });
    const sectionChange = (value) => setData({ ...data, section: value });

    return <Row
      data-testid="filter-form"
      style={{
        justifyContent: "space-around",
      }}
    >
      <Col span={spanSize}>
        <label className="filter__item">
          <span className="filter__item__title">Viral:</span>
          <span className="filter__item__control">
            <Switch
              checkedChildren="including"
              unCheckedChildren="excluding"
              checked={data.showViral}
              onChange={viralChange}
            />
          </span>
        </label>
      </Col>
      <Col span={spanSize}>
        <label className="filter__item">
          <span className="filter__item__title">Window:</span>
          <span className="filter__item__control">
            <Select
              style={{ width: "100%" }}
              placeholder="select window"
              value={data.window}
              onChange={windowChange}
              optionLabelProp="label"
            >
              {renderOptions(imgurFilterQuery.window)}
            </Select>
          </span>
        </label>
      </Col>
      <Col span={spanSize}>
        <label className="filter__item">
          <span className="filter__item__title">Sort:</span>
          <span className="filter__item__control">
            <Select
              style={{ width: "100%" }}
              placeholder="select sort"
              defaultValue={data.sort}
              value={data.sort}
              onChange={sortChange}
              optionLabelProp="label"
            >
              {renderOptions(imgurFilterQuery.sort)}
            </Select>
          </span>
        </label>
      </Col>
      <Col span={spanSize}>
        <label className="filter__item">
          <span className="filter__item__title">Section:</span>
          <span className="filter__item__control">
            <Select
              style={{ width: "100%" }}
              placeholder="select section"
              defaultValue={data.section}
              value={data.section}
              onChange={sectionChange}
              optionLabelProp="label"
            >
              {renderOptions(imgurFilterQuery.section)}
            </Select>
          </span>
        </label>
      </Col>
      <Col span={spanSize} style={{ display: "flex" }}>
        <Button
          disabled={loading}
          className="filter_btn"
          type="primary"
          shape="round"
          data-testid="filter-btn"
          onClick={doFilter}
        >
          <FilterOutlined />
          filter
        </Button>
        {isThereAnyFilter && <Button
          disabled={loading}
          className="clear-filter_btn"
          type="primary"
          shape="round"
          data-testid="clear-filter-btn"
          onClick={restFilter}
        >clear</Button>}
      </Col>
    </Row>
  }, [isThereAnyFilter, loading, data]);
  
  return isMobieOrTablet ? (
    <>
      <FilterOutlined
        onClick={setModalBeVisible}
        className="filter-icon-btn"
        data-testid="filter-btn-icon"
      />
      <Modal
        title="Filter form"
        centered
        footer={false}
        visible={modalVisible}
        onOk={setModalInVisible}
        onCancel={setModalInVisible}
      >
        <StyledDiv className="filter--vertical">
          <Vertical>{filterForm(24)} </Vertical>
        </StyledDiv>
      </Modal>
    </>
  ) : (
      <StyledDiv className="filter ">{filterForm()}</StyledDiv>
    );
};

export default Filter;
