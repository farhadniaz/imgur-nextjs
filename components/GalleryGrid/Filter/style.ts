import styled from "styled-components";

const StyledDiv = styled.div`
  .filter {
    &__item {
      min-height: 32px;
      display: flex;
      &__title {
        padding-right: 8px;
        margin: auto;
      }
      &__control {
        margin: auto;
        width: 100%;
      }
    }
    &_btn {
      width: 100%;
      max-width: 150px;
      margin: 0 auto;
      display: block;
    }
  }
  .clear-filter_btn {
    margin-left: 8px;
  }
`;

export const Vertical = styled.div`
  padding: 24px;
  .filter {
    &__item {
      margin-bottom: 16px;
      &__title {
        width: 100px;
      }
    }
  }
`;

export default StyledDiv;
