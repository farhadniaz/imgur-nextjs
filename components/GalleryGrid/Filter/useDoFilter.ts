import { getGalleryImages } from "~/services/imgur";
import { useStore } from "react-redux";

import GalleryActions from "~/store/gallery/actions.ts";
const useDoFilter = (setLoading) => {
  const store = useStore();
  return (filterData) => {
    setLoading(true);
    getGalleryImages(filterData).then(({ page, data }) => {
      store.dispatch({
        type: GalleryActions.SET_ITEMS,
        payload: data,
      });
      store.dispatch({
        type: GalleryActions.SET_FILTER,
        payload: filterData,
      });
      setLoading(false)
    });
  };
};

export default useDoFilter;
