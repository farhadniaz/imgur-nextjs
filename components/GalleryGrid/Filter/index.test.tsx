import "~/__mocks__/matchMedia.js";
import { render, fireEvent } from "@testing-library/react";
import Filter from "./";
import { Provider } from "react-redux";
import { makeStore } from "~/store/index.ts";
import AppActions from "~/store/app/actions";

const store = makeStore({});

test("Check filter form", async () => {
  store.dispatch({
    type: AppActions.SET_DEVICE,
    payload: { type: "tablet" },
  });

  const { getByTestId } = render(
    <Provider store={store}>
      <Filter onFilter={() => { }} loading={false} />
    </Provider>
  );

  const filterButtonIcon = getByTestId("filter-btn-icon");
  fireEvent.click(filterButtonIcon);

  const filterForm = getByTestId("filter-form");
  fireEvent.click(filterForm);

  const filterButton = getByTestId("filter-btn");

  fireEvent.click(filterButton);
});

test("Check Snapshot in tablet", async () => {
  store.dispatch({
    type: AppActions.SET_DEVICE,
    payload: { type: "tablet" },
  });
  const { asFragment } = render(
    <Provider store={store}>
      <Filter onFilter={() => { }} loading={false} />
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});

test("Check Snapshot in Desktop", async () => {
  store.dispatch({
    type: AppActions.SET_DEVICE,
    payload: { type: "desktop" },
  });
  const { getByTestId, asFragment } = render(
    <Provider store={store}>
      <Filter onFilter={() => { }} loading={false} />
    </Provider>
  );
  const filterForm = getByTestId("filter-form");
  expect(asFragment()).toMatchSnapshot();
});
