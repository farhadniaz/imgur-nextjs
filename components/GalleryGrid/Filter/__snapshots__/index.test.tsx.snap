// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Snapshot in Desktop 1`] = `
<DocumentFragment>
  <div
    class="style__StyledDiv-sc-1032p61-0 fXppqO filter "
  >
    <div
      class="ant-row"
      data-testid="filter-form"
      style="justify-content: space-around;"
    >
      <div
        class="ant-col ant-col-4"
      >
        <label
          class="filter__item"
        >
          <span
            class="filter__item__title"
          >
            Viral:
          </span>
          <span
            class="filter__item__control"
          >
            <button
              aria-checked="false"
              class="ant-switch"
              role="switch"
              type="button"
            >
              <div
                class="ant-switch-handle"
              />
              <span
                class="ant-switch-inner"
              >
                excluding
              </span>
            </button>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-col-4"
      >
        <label
          class="filter__item"
        >
          <span
            class="filter__item__title"
          >
            Window:
          </span>
          <span
            class="filter__item__control"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              style="width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="day"
                >
                  day
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-col-4"
      >
        <label
          class="filter__item"
        >
          <span
            class="filter__item__title"
          >
            Sort:
          </span>
          <span
            class="filter__item__control"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              style="width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="viral"
                >
                  viral
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-col-4"
      >
        <label
          class="filter__item"
        >
          <span
            class="filter__item__title"
          >
            Section:
          </span>
          <span
            class="filter__item__control"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              style="width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                    aria-autocomplete="list"
                    aria-controls="rc_select_TEST_OR_SSR_list"
                    aria-haspopup="listbox"
                    aria-owns="rc_select_TEST_OR_SSR_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="rc_select_TEST_OR_SSR"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="hot"
                >
                  hot
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-col-4"
        style="display: flex;"
      >
        <button
          class="ant-btn ant-btn-primary ant-btn-round filter_btn"
          data-testid="filter-btn"
          type="button"
        >
          <span
            aria-label="filter"
            class="anticon anticon-filter"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="filter"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M880.1 154H143.9c-24.5 0-39.8 26.7-27.5 48L349 597.4V838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V597.4L907.7 202c12.2-21.3-3.1-48-27.6-48zM603.4 798H420.6V642h182.9v156zm9.6-236.6l-9.5 16.6h-183l-9.5-16.6L212.7 226h598.6L613 561.4z"
              />
            </svg>
          </span>
          <span>
            filter
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Check Snapshot in tablet 1`] = `
<DocumentFragment>
  <span
    aria-label="filter"
    class="anticon anticon-filter filter-icon-btn"
    data-testid="filter-btn-icon"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      data-icon="filter"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M880.1 154H143.9c-24.5 0-39.8 26.7-27.5 48L349 597.4V838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V597.4L907.7 202c12.2-21.3-3.1-48-27.6-48zM603.4 798H420.6V642h182.9v156zm9.6-236.6l-9.5 16.6h-183l-9.5-16.6L212.7 226h598.6L613 561.4z"
      />
    </svg>
  </span>
</DocumentFragment>
`;
